crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//! # use size_of::SizeOf;
+//! # use time::Date;
+//!
+//! # #[derive(Clone, Debug, Deserialize, Eq, PartialEq, Ord, PartialOrd, Hash, SizeOf)]
+//! # struct Record {
+//! #     location: String,
+//! #     date: Date,
+//! #     daily_vaccinations: Option<u64>,
+//! # }
+//!
+//! fn build_circuit(circuit: &mut RootCircuit) -> Result<CollectionHandle<Record, isize>> {
+//!     let (input_stream, input_handle) = circuit.add_input_zset::<Record, isize>();
+//!     input_stream
+//!         .filter(|r| {
+//!             r.location == "England"
+//!                 || r.location == "Northern Ireland"
+//!                 || r.location == "Scotland"
+//!                 || r.location == "Wales"
+//!         })
+//!         .inspect(move |record: &OrdZSet<Record, isize>| {
+//!             println!("{}", record.weighted_count());
+//!         });
+//!     Ok(input_handle)
+//! }
+//! ```
+//!
+//! This prints `3083`, showing that we did select a subset of the 3,961
+//! total records.
+//!
+//! The full program is in `tutorial4.rs`.
+//!
+//! ### Aggregating by month
+//!
+//! 3,083 records is still a lot to look at.  There's so much because we've got
+//! years of daily data.  Let's aggregate daily vaccinations into months, to get
+//! monthly vaccinations.  DBSP has several forms of aggregation.  All of them
+//! work with "indexed Z-sets" ([`IndexedZSet`]), which are
+//! Z-sets of key-value pairs, that is, they associate key-value pairs with
+//! weights. Aggregation happens across records with the same key.
+//!
+//! To aggregate daily vaccinations over months by location, we need to
+//! transform our Z-set into an indexed Z-set where the key (the index) has the
+//! form `(location, year, month)` and the value is daily vaccinations (we could
+//! keep the whole record but we'd just throw away most of it later).
+//! To do this, we call [`Stream::index_with`], passing in a function that maps
+//! a record into a key-value tuple:
+//!
+//! ```ignore
+//!     let monthly_totals = input_stream
+//!         .filter(|r| {
+//!             r.location == "England"
+//!                 || r.location == "Northern Ireland"
+//!                 || r.location == "Scotland"
+//!                 || r.location == "Wales"
+//!         })
+//!         .index_with(|r| {
+//!             (
+//!                 (r.location.clone(), r.date.year(), r.date.month() as u8),
+//!                 r.daily_vaccinations.unwrap_or(0),
+//!             )
+//!         })
+//! ```
+//!
+//! Then we can call [`Stream::aggregate_linear`], the simplest form of
+//! aggregation in DBSP, to sum across months.  This function sums the output of
+//! a function of key-value pairs across each unique key.  To get monthly
+//! vaccinations, we just sum the values from our indexed Z-set (we have to
+//! convert to `isize` because aggregation implicitly multiplies by record
+//! weights):
+//!
+//! ```ignore
+//!         .aggregate_linear(|(_c, _y, _m), &v| v as isize)
```

Comment:
why _c and not _l?

Reply:
why does one argument have & and the other doesn't?

Reply:
> why _c and not _l?

At some point I mentally switched from "location" to "country", but you're right. This was all over. I've changed them all to `l` now.

> why does one argument have & and the other doesn't?

Well, they're all borrowed, but we only care about the type of `v` since we ignore the rest. I'll change it to be consistent.